src/controllers/balancesController.ts

```diff
@@ -0,0 +1,119 @@
+import { JsonController, Get, Param, QueryParam, BadRequestError, Post, Delete, HttpCode, OnNull, OnUndefined, Res, Ctx } from "routing-controllers";
+import { BalanceRepository, BalanceEntity } from "../domain/balances";
+import { AssetRepository } from "../domain/assets";
+import { EosService } from "../services/eosService";
+import { ConflictError } from "../errors/conflictError";
+import { BlockchainError } from "../errors/blockchainError";
+
+@JsonController("/balances")
+export class BalancesController {
+
+    constructor(
+        private assetRepository: AssetRepository,
+        private balanceRepository: BalanceRepository,
+        private eosService: EosService) {
+    }
+
+    @Get()
+    async balances(
+        @QueryParam("take", { required: true }) take: number,
+        @QueryParam("continuation") continuation?: string) {
+
+        if (Number.isNaN(take) || !Number.isInteger(take) || take <= 0) {
+            throw new BlockchainError({ status: 400, message: "Query parameter [take] is invalid, must be positive integer" });
+        }
+
+        if (!!continuation && !this.balanceRepository.validateContinuation(continuation)) {
+            throw new BlockchainError({ status: 400, message: "Query parameter [continuation] is invalid" });
+        }
+
+        const block = (await this.eosService.getLastIrreversibleBlockNumber()) * 10;
+        let items: any[] = [];
+
+        // CosmosDB doesn't suppport multiple $match-es in public preview version,
+        // so we can't filter out zero balances on server.
+        // Instead we have to set non-zero balances incrementally
+        // to return exactly [take] number of items
+
+        do {
+            const result = await this.balanceRepository.get(take, continuation);
+
+            continuation = result.continuation;
+
+            for (const e of result.items) {
+                if (e.AmountInBaseUnit > 0) {
+                    items.push({
+                        address: e._id.Address,
+                        assetId: e._id.AssetId,
+                        balance: e.AmountInBaseUnit.toFixed(),
+                        block: Math.max(e.Block, block)
+                    });
+                    take--;
+                }
+            }
+
+        } while (take > 0 && !!continuation)
+
+        return {
+            continuation,
+            items
+        };
+    }
+
+    @Get("/:address/:assetId")
+    async balanceOf(
+        @Param("address") address: string,
+        @Param("assetId") assetId: string) {
+
+        if (!this.eosService.validate(address)) {
```

Comment:
Same here - not sure simple this can be done in js, but it will be great if there is a simple way to move all validations that are used multiple times to separate class. This will allow to do not copy paste errors texts. Can be done later